### PR TITLE
[Ecosystem] [3/3] Add token metadata to indexer

### PR DESCRIPTION
**Pull request by areshand**
## Motivation

1. fetch arweave URI and parse the json to store the metadata in DB
2. add a arg to indexer cli to turn on/off the token indexer. 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md#pull-requests)?
Yes

## Test Plan

Token metadata uri https://xdc7mlwypvsqdrbp3fq2jywgiapdod5fck733ilywdvo3slng3uq.arweave.net/uMX2Lth9ZQHEL9lhpOLGQB43D6USv72heLDq7cltNuk

The token metadata data is stored at DB as expected.
Screen Shot 2022-05-30 at 4 41 57 PM

### CODE REVIEW

File: ecosystem/indexer/src/indexer/metadata_fetcher.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    fn parse_json(&self, value: serde_json::Value) -> Result<TokenMetaFromURI> {
+        Ok(serde_json::value::from_value::<TokenMetaFromURI>(value)?)
+    }
+
+    pub async fn get_metadata(&self, uri: String) -> Option<TokenMetaFromURI> {
+        match get_type(uri) {
+            UriType::ARWEAVE { uri } => match self.read_http_uri(uri).await {
+                Ok(value) => self.parse_json(value).ok(),
+                _ => None,
```
[the remark this answers is not in the document]
Comment (areshand) [reply ecosystem/indexer/src/indexer/metadata_fetcher.rs:58]: added 3 times retry to the metadata fetcher

File: ecosystem/indexer/src/models/metadata.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct TokenMetaFromURI {
+    pub name: Option<String>,
+    pub symbol: Option<String>,
+    pub seller_fee_basis_points: Option<i64>,
+    pub description: Option<String>,
+    pub image: Option<String>,
+    pub external_url: Option<String>,
```
[the remark this answers is not in the document]
Comment (areshand) [reply ecosystem/indexer/src/models/metadata.rs:55]: added

File: ecosystem/indexer/src/schema.rs [hunk 3]
```diff
@@ -39,6 +39,22 @@ table! {
     }
 }
 
+table! {
+    metadatas (token_id) {
+        token_id -> Varchar,
+        name -> Nullable<Varchar>,
+        symbol -> Nullable<Varchar>,
+        seller_fee_basis_points -> Nullable<Int8>,
```
[the remark this answers is not in the document]
Comment (areshand) [reply ecosystem/indexer/src/schema.rs:47]: what is the comment ? I didn't find it

File: ecosystem/indexer/src/token_processor.rs [hunk 4]
```diff
@@ -58,6 +59,19 @@ fn update_mint_token(conn: &PgPoolConnection, event_data: MintEventType, txn: &U
     query.execute(conn).expect("Error updating row in token");
 }
 
+async fn get_all_metadata(uris: &Vec<(String, String)>, res: &mut Vec<Metadata>) {
+    let fetcher = MetaDataFetcher::new();
+    for (tid, uri) in uris {
+        let token_metadata = fetcher.get_metadata(uri.clone()).await;
```
[the remark this answers is not in the document]
Comment (areshand) [reply ecosystem/indexer/src/token_processor.rs:65]: maybe leave this for future improvement now.

File: ecosystem/indexer/src/schema.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hariria) [reply ecosystem/indexer/src/schema.rs:47]: Oh i just meant add the animation_url here, you got it

**Review by hariria — APPROVED** [review]
LGTM

**Pull request closed without merging**